### ISSUE TITLE
64-bit int support in MFEM?

### ISSUE THREAD

**Issue by najlkin** [mfem/mfem#1515]
Hi,
it is a matter of fact that MFEM uses plain `int` throughout the code, which are 32-bit in GCC even on 64-bit processors. It can be compiled against HYPRE or PETSc with 64-bit integers, but these integers are used only for global indexing and the local interface remains with 32-bit integers. This restricts the size of all arrays and matrices to 2^31elements, which can be translated for doubles to 16 GB of memory.
For example, the library is a part of the exascale project, so suppose that you want to use 1 exabyte array. A quick math says that you would need about whopping 64 million processes.
Is there a slight chance that MFEM will use 64-bit integers one day? The setting can be centralized similarly to HYPRE, so there could be typedef of something like `MFEM_Int`. I know it is relatively late for that and it would require extensive changes in the code, but it is never too late, right? :smile:

**Comment by tzanio** [comment]
Hi @najlkin,

Switching all ints to 64-bit with something like `MFEM_Int` will likely slow down the code, so you probably want to do that selectively. 

Our current approach is that on each MPI task the number of unknowns is at most in the tens of millions, and so locally we only need 32-bit ints (globally of course we do need the 64-bit).

I think this is actually quite reasonable given the current memory per node and strong scaling considerations on parallel machines.

Do you have an use case where you need more than 2B unknowns in an MPI task?

Tzanio

**Comment by v-dobrev** [comment]
In the context of finite elements, 32bits is currently and for the foreseeable future more than sufficient. Look at it this way: both Summit and Sierra have 16GB per GPU (which is typically = 1 MPI rank) and you definitely need more than 1 vector to do anything of interest. On Fugaku, according to https://postk-web.r-ccs.riken.jp/spec.html, you will have 32GB/node where each node has 1 CPU with 48 cores, so 32bit ints are still more than sufficient.

**Comment by najlkin** [comment]
It does need to be really all integers, but basically all used for computing like indexing, DOFs and memory related operations, where the standard library uses type `size_t` for this purpose anyway, which is 64-bit on 64-bit machines. More granular setting would be nice, but it is clear it is not manageable and would cause myriad of unforeseen problems.

It is true that the current trend is rather opposite, where the hardware development of memory stays behind the one of cpus, but relaying only on 32-bit variables seems to me obsolete and dangerous. Overflows happen easily and stay under the radar. I noticed one just when after a long debugging session when I found out that a matrix actually tries to allocate a negative amount of memory.

My use case was that I got hands on a machine with 48 physical cores and 1.5 TB operating memory. I thought to utilize rather OpenMP, but even with MPI its hard use whole memory with that limit of 16 GB.

**Comment by tzanio** [comment]
Hi Jan,

We discussed this today and while it will be nice to support size_t where necessary, the consensus was that adding such support will be a significant amount of work.

It is also not clear to me how necessary this work really is.

Can you try using MPI for your problem? This much better supported than OpenMP in mfem, and if need be you can use more MPI ranks than cores to solve really big problems. 

To give us an idea -- what is the global problem size you are targeting, and at what local problem size per MPI task do you start seeing problems?

Thanks for bringing this up, it's a good question.
Tzanio

**Comment by najlkin** [comment]
I have a little bit hungry 4D (+time) code and I am trying to make convergence analysis, so basically: the more DOFs, the better. When I ran into the problems, I had about 10 million DOFs locally, so `int` was totally fine indexing DOFs, but the size of the sparse matrix already overflowed. To give you an idea, I had just 13 x 13 x 13 cubic elements locally x 40 linear in that extra dimension (it is the vector dimension of the elements in MFEM). With higher order elements I can afford only even less (in terms of DOFs), because the sparsity is lower.

However, I know that my application is a bit specific and I need to invest in more MPI processes (so you may close this, if you want), but you have what to think about at least :smile:.

**Comment by v-dobrev** [comment]
Is there something preventing you from using 48 or even 96 MPI tasks on the machine with 1.5TB?

The SparseMatrix class is maybe one place where using 64bit indexing may make sense. Specifically, the offset array `I` can hold 64bit offset indices while the column-indices array, `J`, can still use 32bits entries (while it's size will be 64bit -- indexing into `J` goes through `I`). This way you can have matrices with dimensions of up to 2 billion rows/columns but the number of nonzeros can be much larger. The size of the `I` array (= number of rows + 1) is typically much smaller than the size of the `J` and `data` arrays (= number of nonzeros), so just switching `I` to 64bit will have small effect on the overall memory usage (and speed) of the SparseMatrix class.

In your case, such a change will resolve the issue you ran into, right?

**Comment by najlkin** [comment]
Yes, you are right. The offset array `I` is the place where 64bit integers are most desperately needed. However, the change would involve not only a single memory vector, but all operations accessing the data on low level, counting or indexing the nonzero elements and `Memory` class handling the memory allocation. Moreover, `Memory` is exactly the place where `size_t` should appear anyway and it does partially, as it is used in `MemoryManager`, which is a bit inconsistent at the moment.

**Comment by termi-official** [comment]
Hi guys,

I think I should hook into this discussion, as I am not sure if you are aware that signed integers and unsigned integers (also note, size_t is an unsigned integer) have different semantics in C++. While signed integer over-/underflow is undefined behavior, unsigned integer under-/overflow has wrap-around semantics. Perhaps more intuitively, signed integers work as expected (from the whole numbers), while unsigned integers behave like a modulo ring. Using unsigned integers can prevent optimizations in general and lead to really strange, hard to track bugs. 

To give a simple example, the expression x < x+1 is always true for signed integers, but not for unsigned integers (because the largest unsigned integer + 1 = 0).

The assembly for 
```cpp
using T = unsigned int;
int main(int argc, const char** argv)
{
    T cond = static_cast<T>(argc);
    if(cond < cond + 1)
    {
        return 1;
    }
    return 0;
}
```
with clang (trunk) and -O3 is
```
main:                                   # @main
        xor     eax, eax
        cmp     edi, -1
        setne   al
        ret
```
and now switching to ``T = int`` it changes the assembly
```
main:                                   # @main
        mov     eax, 1
        ret
```
eliminating the remainder of the branch.

``int`` has only the guarantee to be at least 16 bit wide. So, if you want to go down the route of large indices I would suggest to use ``int64_t``, which has the guarantee to be exactly 64 bit wide.

Best regards,
Dennis

**Comment by v-dobrev** [comment]
@najlkin, yes, even if we want to update just the `SparseMatrix` class (and its Boolean version, class `Table` with it), the changes will need to propagate to other places in the code. It will be best to support this only when a new build option is enabled, e.g. something like `MFEM_USE_BIG_NNZ`. That way any interface changes in the code will only affect users that enable that option. Or maybe we can avoid changing the interface (performing copies when conversion is needed) and just extending the interface for the case of 64bit `int`s in the `I` array. In the `Memory` class we can simply introduce a second template parameter (with default value of `int`) for the indexing/size type.

I think such a set of changes will be much more limited compared to converting all `int`s to one single `typedef`ed MFEM type. To really see the full extend of the necessary changes, however, we'll need to try this out.

**Comment by najlkin** [comment]
Hi @termi-official, thanks for your theoretical contribution. I would need to search the standards, but I am convinced that simple binary complement logic works on most of the machines, so `int` and `unsigned` behave the same way on the low level. The code:
```C++
#include <iostream>

using namespace std;

int main()
{
	int a = ~0u >> 1;
	cout << a << endl;
	int b = a + 1;
	cout << b << endl;
	cout << boolalpha << (a < a+1) << endl;
	cout << boolalpha << (a < b) << endl;
}
```
gives me:
```
2147483647
-2147483648
true
false
```
so it is just the condition that is interpreted differently. Isn't it better to use something that has well-defined behavior and consistent logic then? Exactly this kind of condition helps to detect overflows and avoid them. Besides that, the conversion is performed by the compiler anyway, because `Memory` uses `size_t` internally (and it has to, because C++ standard library uses the type for memory-related stuff). However, most of the functions in `Memory` are inlined, so `size_t` is propagated to user code anyway and it is on the compiler to decide how to optimize the thing.

We do not need exactly 64-bit integers, but the "native" integer, because we are talking about memory allocation. If you have 32-bit machine (or 16-bit, but hey, you should really buy a newer one :smile: ) then you cannot operate on a bigger memory space than that (ok, there used to be some hacks in the old days, but that is not relevant). Therefore, the question is just when or if it is reasonable to use a smaller than the native integer like `int` in GCC.

@v-dobrev , as I see it now, I agree that it makes most sense to limit the scope just to `SparseMatrix` and similar things and only optionally probably, so there could be just a little adorable typedef locally :smile:. About `Memory` class I am not sure, using that template parameter also came to my mind as the most straightforward thing. On the other hand, `size_t` is used there anyway and the only performance hit would be loading 64-bit instead of 32-bit `capacity` variable from memory. As it is just a single variable and it is used frequently, I think it is cached anyway, so there might be next to zero performance impact. Using a single fixed implementation is safer from the development point of view and also switching between signed and unsigned types in an uncontrolled manner through the template parameter is not very clean.

**Comment by termi-official** [comment]
Thanks for the reply @najlkin . You are right, on the usual architectures (i.e. the server and desktop Intel and AMD ones) this is exactly what happens low-level. We should be aware, that this is not guaranteed to happen in general, especially since the trend in computing goes towards accelerator hardware. I think I was a imprecise on my actual point here, sorry.

In C++ we make a trade-off between defining behavior and performance. Hitting undefined behavior can also be interpreted as programmatic errors. In code optimization we can make the simplifying assumption, that the program to be optimized is error-free, so undefined behavior becomes "things that cannot happen". If something cannot happen, then we are free to remove the associated code and/or skip possible checks. This is extremely useful for making a program fast, but comes at the expense of throwing away the general well-definiteness. In the case of unsigned vs signed when indexing things, this **may** hurt performance in a severe way, if the unsigned type gets propagated to the "important loops" in our programs, possibly introduction jumps and also possibly preventing loop transformations like vectorization (which for example happens when the compiler fails to prove that the wrap-around is impossible).

I totally agree with you, that for this behavior is a nightmare for debugging and well-defined behavior helps in this point a lot (although I do not see how it helps avoiding these errors). For me this is a problem which can be solved with tooling, without sacrificing performance in general. Especially tooling for C++ improved a lot through the last years, if you take a look at the newly developed sanitizers and static analyzers. Also note that, although the code with size_t is inlined, the type does not propagate through the user code.

@v-dobrev I think having the option to use "big integers" indexing would be helpful for computations which require just one node (i.e. shared memory parallelism), since some implementations (e.g. when intermediate computations with non-local quantities arise) get simpler. To give a specific example, I am working on a high-dimensional (dynamical) reaction-diffusion problem in electrophysiology involving a bidirectional 1D to 3D coupling and adaptive mesh refinement. The coupling must be updated after each AMR step (because the elements which are coupled may be de-/refined), but rebalancing also may move elements around the processes, introducing nasty corner cases in a distributed implementation. 
Now, to efficiently solve my reaction-diffusion I use the default operator splitting scheme (separating reaction and diffusion and solve them alternatingly). The AMR may exceed 10mio elements for several timesteps, which can get problematic as the total number of DOFs in these timesteps can exceed the ~2 billion mark for 32 bit integers (because in several problems the reaction is >100 dimensional). I know, this is a quite exotic usage of finite elements and the option is just nice to have. 

Maybe this could be a good first issue for someone who wants to get involved?

**Comment by v-dobrev** [comment]
@termi-official, I guess your argument for big integer indexing (for things like number of mesh elements, number of dofs, etc) is the following: if a user is restricted to running in "serial" (i.e. no MPI parallelism), potentially due to factors outside of MFEM, then the 32bit barrier of 2 billion dofs (or mesh entities like elements, faces, edges, vertices) can be a problem, if the compute node has a lot of memory. Did I understand you correctly?

This is not a typical HPC system and use case. Also, the first barrier such a case will hit (unless using matrix-free algorithms) is the number of non-zeroes in your `SparseMatrix` objects. These are typically at least an order of magnitude larger than the number of dofs. That's why I'm suggesting here to address that barrier first. If that turns out to be insufficient in some practical case, we can look into the "dof barrier" which will require more extensive changes.

**Comment by termi-official** [comment]
Yes, this is correct. I also absolutely agree, that this is really not a typical HPC use case. I just wanted to point out, that practical scenarios exists, in which 64bit DOF indices can come in handy. :)

**Comment by sfmahenk** [comment]
Hi,

I am rather new to MFEM and I have a similar setup to @najlkin (single workstation with 512GB RAM). I naively tried to display a mesh (ca. 1e8 nodes, ca. 6e8 tets, file format: gmsh version 2 binary) with GLVis.

This crashes: 
```
[...]
terminate called after throwing an instance of 'std::bad_array_new_length'
  what():  std::bad_array_new_length
  
Program received signal SIGABRT, Aborted.
0x00002aaaac758428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
```
The backtrace of the stack reveals:
```
(gdb) backtrace 
#0  0x00002aaaac758428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00002aaaac75a02a in __GI_abort () at abort.c:89
#2  0x00002aaaabeb8dde in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00002aaaabec4896 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00002aaaabec4901 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00002aaaabec4b55 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00002aaaabeb8946 in __cxa_throw_bad_array_new_length () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x000000000041a7b8 in mfem::Memory<int>::New (this=0xae0800, size=-460808932) at ../mfem-4.1/general/mem_manager.hpp:628
#8  0x000000000049000c in mfem::Table::Table (this=0xae07e0, dim=639026394, connections_per_row=6) at general/table.cpp:60
#9  0x00000000004d20d6 in mfem::Mesh::GetElementToFaceTable (this=0xae3070, ret_ftbl=0) at mesh/mesh.cpp:5146
#10 0x00000000004c20c1 in mfem::Mesh::FinalizeTopology (this=0xae3070, generate_bdr=true) at mesh/mesh.cpp:2265
#11 0x00000000004c8334 in mfem::Mesh::Loader (this=0xae3070, input=..., generate_edges=1, parse_tag="") at mesh/mesh.cpp:3302
#12 0x00000000004f2a14 in mfem::Mesh::Load (this=0xae3070, input=..., generate_edges=1, refine=0, fix_orientation=false) at mesh/mesh.hpp:670
#13 0x00000000004c633c in mfem::Mesh::Mesh (this=0xae3070, input=..., generate_edges=1, refine=0, fix_orientation=false) at mesh/mesh.cpp:2995
#14 0x00000000004110e7 in ReadSerial () at glvis.cpp:1756
#15 0x000000000041085f in main (argc=3, argv=0x7fffffffe0a8) at glvis.cpp:1584
```
So the cause is an integer overflow (639026394 * 6 = -460808932) and this is how I found this github issue :)

My current strategy is based on ex1p.cpp. I generate a coarser version of the mesh with gmsh, load this coarse mesh serially with MFEM on N MPI processes, partition and refine in parallel to achieve the desired mesh resolution and finally solve the problem.

Similar to the example I save mesh and solution in MFEM's parallel format. Unfortunately I have difficulties to visualize the solution. GLVis crashes when trying to merge the parallel mesh.

Further limitation of this strategy: Since each of the N MPI processes loads the coarse mesh, one runs easily into memory limitations and I cannot utilize the CPU to full capacity. This is closely related to issue #1540

I would be very grateful if any of you could point out to me how to approach problems of this size with MFEM.

**Comment by sshiraiwa** [comment]
Regarding the last point, #934 had some good suggetions

**Comment by v-dobrev** [comment]
Hi @sfmahenk,

Thank you for letting us know about the issues you ran into.

Since you seem to have a big problem, I'd definitely recommend solving it using all cores on your machine -- the best option for that in mfem is to use the MPI parallelism.

Regarding the GLVis issue with large meshes -- this is inherent problem in GLVis and the way it uses OpenGL. It basically tries to load all the elements (all surface elements in 3D) in an OpenGL display list which are stored in video memory. This puts even more stress on your video card which typically has a lot less memory than the CPU memory.

So for this reason, for big meshes, I'd recommend using VisIt (or ParaView) -- these were meant to handle large parallel meshes. Take a look at `ex5p.cpp` -- it illustrates how to use the `VisItDataCollection` and `ParaViewDataCollection` classes.

Regarding the issue of loading the "coarse" mesh on all MPI ranks: I see how this can be an issue if the "coarse" mesh is not that coarse. For example, if your coarse mesh is 8 times smaller than your final "fine" mesh and you are using 16 MPI ranks, at the beginning the total memory requirement is `16*coarse_size ~ 2*fine_size` which is not very helpful. In your case, how much coarser is the "coarse" mesh compared to the "fine" mesh? Is there a reason you cannot use even coarser mesh as your starting "coarse" mesh, e.g. something that is `8^2` or even `8^3` times smaller than the final "fine" mesh you want to solve on?

If starting with an even coarser mesh is not a good option for you, we can try to prioritize this issue even more -- it has already come up a number of times -- and implement a reasonable solution for it.
